### PR TITLE
Migrate Transform

### LINKED ISSUE #111: Support Batch-Level and Sample-Level Transformations
# Description

Currently, medicai's transformations are primarily designed to operate on individual samples. As the [pre-requisite](https://github.com/innat/medic-ai/issues/104) on batch level data processing and GPU-augmentation, it would be useful to support transformations that can operate directly on batched tensors as well.

The goal is to make transformations **batch-aware** by design, while allowing individual sample inputs when appropriate. Ideally, transformations should support one of the following modes depending on the nature of the operation:

- **Sample-level** only: when the operation is inherently sample-specific, i.e. `Spacing`, `CropForeground`, etc.
- **Dual-mode** (Sample-level and Batch-level): when the same transformation can naturally operate on both sample and batch inputs.

For dual-mode transformations, the implementation should preferably have a single batch-oriented internal implementation. When a sample-level tensor is provided, the transform can temporarily add a dummy batch dimension, execute the batch implementation, and remove the dummy dimension before returning the result.

# Motivation

Supporting batch-level transformations can provide several benefits:

- Avoid maintaining separate implementations for sample and batch operations.
- Make transformations easier to integrate with batched data pipelines.
- Enable more efficient vectorized operations.
- Allow transformations to take advantage of backend-specific batch operations where available, i.e. GPU augmentation.

# Example

A spatial transformation such as `Rotate90` could support:

```python
Sample:
(H, W, C)
(D, H, W, C)

Batch:
(B, H, W, C)
(B, D, H, W, C)
```

with the actual rotation implementation operating only on the batch representation.

# Usages

**Sample-only transform**

```python
import tensorflow as tf
from medicai.transforms import Spacing

spacing = Spacing(
    keys=["image", "label"],
    pixdim=(2.0, 1.5, 1.5),
    interpolation=("trilinear", "nearest"),
    input_layout="DHWC",
)

sample = {
    "image": tf.random.normal((96, 128, 128, 1)),
    "label": tf.random.uniform((96, 128, 128, 1), maxval=2, dtype=tf.int32),
}
meta = {"affine": tf.eye(4)}

result = spacing(sample, meta)
```

**Dual-mode transform in sample pipeline**

```python
import tensorflow as tf
from medicai.transforms import Compose, CropForeground, Resize

pipeline = Compose(
    [
        CropForeground(
            keys=["image", "label"], source_key="image", input_layout="DHWC"
        ),
        Resize(
            keys=["image", "label"],
            target_shape=(128, 128, 128),
            interpolation=("trilinear", "nearest"),
            input_layout="DHWC",
        ),
    ]
)

sample = {
    "image": tf.random.normal((96, 128, 128, 1)),
    "label": tf.random.uniform((96, 128, 128, 1), maxval=2, dtype=tf.int32),
}

result = pipeline(sample)
```

**Dual-mode transform on a batch**

```python
import tensorflow as tf
from medicai.transforms import Resize

resize = Resize(
    keys=["image"],
    target_shape=(128, 128),
    interpolation="bilinear",
    input_layout="BHWC",
)

batch = {
    "image": tf.random.normal((8, 256, 256, 1)),
}

result = resize(batch)
```

**Dual-mode transform inside train_step to leverage GPU-augmentation**

```python
import keras
import tensorflow as tf
from medicai.transforms import Flip, Rotate90, ShiftIntensity

flip = Flip(keys=["image"], spatial_axis=1, input_layout="BHWC",)
rotate = Rotate90(keys=["image"], k=1, spatial_axis=(1, 2), input_layout="BHWC",)
shift = ShiftIntensity(keys=["image"], offset=0.1, input_layout="BHWC",)


def augment_batch(x, y):
    x = flip({"image": x})["image"]
    x = rotate({"image": x})["image"]
    x = shift({"image": x})["image"]
    return x, y


class ExtendedModel(keras.Model):
    def __init__(self, model, **kwargs):
        super().__init__(**kwargs)
        self.model = model

    def call(self, x, training=False):
        return self.model(x, training=training)

    def train_step(self, data):
        x, y = data
        x, y = augment_batch(x, y)
        return super().train_step((x, y))
```

### PR DESCRIPTION
**Pull request by innat**
fix: https://github.com/innat/medic-ai/issues/111 https://github.com/innat/medic-ai/issues/104

## Summary by CodeRabbit

- **New Features**
  - Transform pipelines now support TensorFlow, JAX, and Torch with backend-neutral operations.
  - Added explicit channel-last layouts for 2D/3D sample and batch tensors.
  - Added seeded random transforms with reproducible behavior and shared batch decisions.
  - Added affine-aware resampling, interpolation, orientation, and spatial transformation capabilities.
  - Added cross-backend training and performance benchmark coverage.

- **Documentation**
  - Expanded guidance for backends, dataloaders, layouts, batching, and transform usage.
  - Added support matrices, examples, and benchmark results.

- **Breaking Changes**
  - Several transforms now require `input_layout`; intensity scaling uses range sequences.
  - Label-aware cutout parameters and selected legacy exports are no longer available.

### CODE REVIEW

**Comment by innat** [comment]
@coderabbitai review

**Comment by innat** [comment]
@coderabbitai full review
